llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1265,6 +1265,120 @@ LogicalResult TopkOp::verify() {
+          builder.create<arith::ConstantIndexOp>(loc, inputShape[dim]);
+    }
+    loopBounds[dim].stride = one;
+  }
+  return loopBounds;
+}
+
+SmallVector<StringRef> TopkOp::getLoopIteratorTypes() {
+  SmallVector<StringRef> iteratorTypes(getInputRank(),
+                                       getParallelIteratorTypeName());
+  iteratorTypes[dimension()] = getReductionIteratorTypeName();
+  return iteratorTypes;
+}
+
+LogicalResult TopkOp::generateScalarImplementation(OpBuilder &b, Location loc,
+                                                   ValueRange ivs) {
+  uint64_t kDim = dimension();
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+  Value initialValue = b.create<memref::LoadOp>(loc, values(), ivs);
+  Value initialIndex = b.create<memref::LoadOp>(loc, indices(), ivs);
+
+  // Compute K (ub) from the selected dim of the output
+  Value ub;
+  ShapedType outputValuesType = outputValues().getType().cast<ShapedType>();
+  if (outputValuesType.isDynamicDim(dimension())) {
+    ub = b.create<memref::DimOp>(loc, outputValues(), dimension());
+  } else {
+    ub = b.create<arith::ConstantIndexOp>(
+        loc, outputValuesType.getDimSize(dimension()));
+  }
+
+  // Inner K loop functions:
+  //   Load current K val/ind
+  //   Compare N to K using inserted block compare
+  //   Check if N/K are equal, which came first
+  //   Store results in appropriate places
+  //   Yield loop carry vals
+  Value kValue, kIndex;
+  auto scfFor = b.create<scf::ForOp>(
+      loc, zero, ub, one, ValueRange{initialValue, initialIndex},
+      [&](OpBuilder &b, Location loc, Value iv, ValueRange loopCarryValues) {
+        SmallVector<Value> indices(ivs);
+        indices[kDim] = iv;
+        kValue = b.create<memref::LoadOp>(loc, outputValues(), indices);
+        kIndex = b.create<memref::LoadOp>(loc, outputIndices(), indices);
+      });
+
+  SmallVector<Value> indices(ivs);
+  indices[kDim] = scfFor.getInductionVar();
+  auto loopCarryValues = scfFor.getRegionIterArgs();
+
+  // Retrieve the comparison from the region and plug into our loop
+  auto &srcBlock = region().front();
+  Region &region = scfFor.getRegion();
+  BlockAndValueMapping bvm;
+  {
+    OpBuilder::InsertionGuard guard(b);
+    auto &block = region.front();
+    b.setInsertionPointToEnd(&block);
+    for (auto it : llvm::zip(srcBlock.getArguments(),
+                             ValueRange{loopCarryValues[0], kValue})) {
+      bvm.map(std::get<0>(it), std::get<1>(it));
+    }
+    for (auto &blockOp : srcBlock.without_terminator()) {
+      b.clone(blockOp, bvm);
+    }
+  }
+  Value valueCmpRes =
+      bvm.lookupOrDefault(srcBlock.getTerminator()->getOperand(0));
+  OpBuilder::InsertionGuard g(b);
+  b.setInsertionPointToEnd(&region.front());
+
+  // Check if the two values are equal, which index came first.
+  Value cmpEqValRes = b.create<arith::CmpFOp>(loc, arith::CmpFPredicate::OEQ,
```

Comment:
If you are using a region to capture the "ordering" then you cannot use a direct comparison of values here. What we are really looking for is really strict weak ordering in the style of C++ comparators.
Effectively the region of the op is a function f(x,y) which returns true to say "x is less than y". It should also return false for f(x, x). Then equality is obtained by checking f(x, y) is false and f(y, x) is false. See for reference https://www.boost.org/sgi/stl/StrictWeakOrdering.html

Also I actually dont see where you are using the body from the `topk` operation here. It seems like the implementation is implicitly assuming you are comparing using standard arithmetic "less than". THe intent of the region is to encode that information in the op region and use that as a "black-box function". We can discuss more on GVC about this.

Reply:
I can try the the strict weak ordering by f(x, y) and  f(y, x). 

For the function call, I copied this implementation from `SortOp` that accepts a comparison block then lifts the comparison out of that block and inserts it into the IR. I don't fully follow how this works entirely using the `BlockAndValueMapping` and `InsertionGuard` semantics, but the result is that is uses the body of the block directly without any function call semantics. Are you suggesting this should not be done that way and if so, how can I maintain the block as a whole?